spec/fixed_price_server_spec.rb

```diff
@@ -1,7 +1,184 @@
+    it "attaches pm, sets default pm, creates sub successfully" do
+      customer_id = Stripe::Customer.create.id
+      pms_before = Stripe::PaymentMethod.list(
+        type: 'card',
+        customer: customer_id
+      )
+      resp, status = post_json("/create-subscription", {
+        paymentMethodId: 'pm_card_visa',
+        customerId: customer_id,
+        priceId: 'BASIC',
+      })
+
+      # It attaches the payment method.
+      pms_after = Stripe::PaymentMethod.list(
+        type: 'card',
+        customer: customer_id
+      )
+      expect(pms_after.data.length).to eq(pms_before.data.length + 1)
+
+      # Sets default payment method
+      customer_after = Stripe::Customer.retrieve(customer_id)
+      expect(customer_after["invoice_settings"]["default_payment_method"]).to eq(
+        pms_after.data.first.id
+      )
+
+      # Creates subscription
+      expect(status).to eq(200)
+      expect(resp).to have_key("id")
+      expect(resp["id"]).to start_with("sub_")
+
+      # Expands latest_invoice.payment_intent
+      expect(resp).to have_key("latest_invoice")
+      expect(resp["latest_invoice"]).to have_key("payment_intent")
+      expect(resp["latest_invoice"]["payment_intent"]).to have_key("client_secret")
+      expect(resp["latest_invoice"]["payment_intent"]["client_secret"]).to start_with("pi_")
+    end
+
+    it "fails to attach pm to customer and returns an error for bad cards" do
+      customer_id = Stripe::Customer.create.id
+      pms_before = Stripe::PaymentMethod.list(
+        type: 'card',
+        customer: customer_id
+      )
+      resp, status = post_json("/create-subscription", {
+        paymentMethodId: 'pm_card_chargeDeclined',
+        customerId: customer_id,
+        priceId: 'BASIC',
+      })
+
+      # It attaches the payment method.
+      pms_after = Stripe::PaymentMethod.list(
+        type: 'card',
+        customer: customer_id
+      )
+      expect(pms_after.data.length).to eq(pms_before.data.length)
+
+      # Sets default payment method
+      customer_after = Stripe::Customer.retrieve(customer_id)
+      expect(customer_after["invoice_settings"]["default_payment_method"]).to be_nil
+
+
+      # Assert error status
+      # TODO: Make this a 400!
+      expect(status).to eq(200)
+      expect(resp).to have_key("error")
+      expect(resp["error"]).to have_key("message")
+      expect(resp["error"]["message"]).to eq("Your card was declined.")
+    end
+  end
+
+  describe "/retrieve-upcoming-invoice" do
+    it "retrieves upcoming invoice for the customer" do
+      customer_id = Stripe::Customer.create.id
+      resp, _ = post_json("/create-subscription", {
+        paymentMethodId: 'pm_card_mastercard',
+        customerId: customer_id,
+        priceId: 'BASIC',
+      })
+      subscription_id = resp["id"]
+      items = resp.dig("items", "data")
+      expect(items.length).to eq(1)
+      old_price = items.first.dig("price", "id")
+
+      # TODO: This should be a GET request.
```

Comment:
Is this meant to still be TODO here?

Reply:
yeah going to come back and clean those up :)